### PR TITLE
CR-1120194: Addressed hang issue for external IO example, specific thread polls the client file descriptor periodically

### PR DESCRIPTION
**Pull request by sgundime-xilinx**
Fix : Currently XRT is hanging once simulator is connected and exits after time. So XRT should be notified if socket is no more active. But XRT cannot except a graceful exit from simulator so a polling mechanism is introduced to track the state change of client socket periodically.

Risk: very Low.
Tests: Canary run. Results are clean.
Reviewer: venkatp

### CODE REVIEW

**Review by venkatp-xilinx — APPROVED** [review]

**Comment by lforg37** [comment]
Wouldn't this PR be the occasion of get rid of the `AQUIRE_MUTEX` and `RELEASE_MUTEX` macro ?
The second one is no more used here, so instead of having an empty body it can be removed everywhere, and maybe it would be time to replace the first one with a more stable construct ?

Because there is a danger that if for some reason the AQUIRE_MUTEX was called in a subscope, the section protected by the mutex changed.

Dummy example code:

```
{
// ...
  {
    // Do some sub-scope local things
    AQUIRE_MUTEX()
    // Other sub-scope local things that needs to be mutex protected
  }
  // Other things that needs to be mutex protected
  RELEASE_MUTEX()
}
```

The other things that needs to be mutex protected will not be anymore with this change.

I am not aying that such code exists in the code base, but keeping the macro name to acquire and release does not communicate the fact that a context manager is created, so could be misleading.

**Comment by kbhardwa-xilinx** [comment]
retest this please

**Comment by dayeh-xilinx** [comment]
kbhardwa-xilinx is re-testing this to verify some infra change.

**Comment by kbhardwa-xilinx** [comment]
retest this please

File: src/runtime_src/core/include/xcl_api_macros.h [hunk 1]
```diff
@@ -37,24 +41,24 @@ mtx.unlock();
     auto c_len = c_msg.ByteSize();                                      \
     buf_size = alloc_void(c_len);                                       \
     bool rv = c_msg.SerializeToArray(buf,c_len);                        \
-    if(rv == false){std::cerr<<"FATAL ERROR:protobuf SerializeToArray failed"<<std::endl;exit(1);} \
+    if(rv == false){std::cerr<<"FATAL ERROR:protobuf SerializeToArray failed"<<std::endl; exit(1);} \
```
[the remark this answers is not in the document]
Comment (sgundime-xilinx) [reply src/runtime_src/core/include/xcl_api_macros.h:44]: The errors were different. So updated with appropriate failures and no new function is introduced to captrure the generic failure in macro.

File: src/runtime_src/core/pcie/emulation/common_em/unix_socket.cxx [hunk 2]
```diff
@@ -93,64 +100,200 @@ void unix_socket::start_server(double timeout_insec, bool fatal_error)
...
+    }
+    if (retval == 0)       
+      continue;                                                 // Poll is timedout so can retry this.
+    
+    // The same logic can be implemented by looking at ~POLLIN or POLLRDHUP & POLLERR & POLLHUP
+    if( mpoll_on_filedescriptor.revents & POLLHUP ) {
+      std::cerr<<"\n Client socket state has changed & it is not readable anymore! So application will exit now.";
+      m_is_socket_live.store(false);
+      server_started.store(false);                              // Let's other monitoring threads exit
+      std::this_thread::sleep_for(5s);                          // provide some time so that spawned threads will terminate themselves messagesThread.
```
[the remark this answers is not in the document]
Comment (sgundime-xilinx) [reply src/runtime_src/core/pcie/emulation/common_em/unix_socket.cxx:283]: The sleep statement is not mandatory to present and it is not a synchronize methodology followed. So removing it. thread.join() would take care of it as you mentioned.

File: src/runtime_src/core/pcie/emulation/common_em/unix_socket.cxx [hunk 3]
```diff
@@ -93,64 +100,200 @@ void unix_socket::start_server(double timeout_insec, bool fatal_error)
...
+
+  while( true ) {
+
+    if ( false == server_started.load() ) {
+      std::cerr<<"\n socket connect is not established/broken \n";
+      break;
+    }
+    
+    mpoll_on_filedescriptor = {fd, POLLERR, 0};                 // monitor client socket file descriptor for a state change.
+    auto retval = poll(&mpoll_on_filedescriptor, 1, 500);       // Let's do polling on it for utmost 500 ms
```
[the remark this answers is not in the document]
Comment (sgundime-xilinx) [reply src/runtime_src/core/pcie/emulation/common_em/unix_socket.cxx:270]: The monitoring functionality requires to be run in parallel and any state change should be notified. So a thread is a good to opt for or std::launch::async would justify the requirement.

1. this poll can be merged with the select in sk_write and sk_read
answer: Agree partially. But what if there is no simulator running and no RPC calls made? The current thread notifies XRT about socket state change and exit the application when next RPC call is made.

File: src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/shim.cxx [hunk 4]
```diff
@@ -1674,9 +1679,11 @@ uint32_t HwEmShim::getAddressSpace (uint32_t topology)
       }
       return;
     }
-
-    resetProgram(false);
-
+    // All RPC calls fail if no socket is live.
+    if (sock->m_is_socket_live != false) {
```
[the remark this answers is not in the document]
Comment (sgundime-xilinx) [reply src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/shim.cxx:1683]: Agreed. updated code accordingly.

**Comment by sgundime-xilinx** [comment]
@Ralender : Can the code be reviewed now please?

**Review by chvamshi-xilinx — APPROVED** [review]

**Review by lforg37 — CHANGES_REQUESTED** [review]
Please see the review comments.

**Comment by lforg37** [comment]
While I can understand that we postpone the remarks related to refactoring, the part of the code related to "non blocking" IO with actively restarting the wait without timings is clearly suspicious, and will probably introduce waste of cpu time.

I don't think merging a PR that adds technical debt, do quite complex modification without a high level comment  explaining that and even keep old code into comment is something we should do. 

Especially in this part of the system which, as has been mentionned by someone else, is already complex enough to be responsible of a good part of the bugs we identified in the emulator.

**Comment by keryell** [comment]
@sonals @stsoe at the end I do not think it is useful to add more buggy code like this PR to XRT.
Perhaps it seems even better to not merge these kinds of random PR and just live with the current situation which does not seem to impact the AMD financial results anyway.
I suggest AMD AECG Research Labs stopping wasting time on reviewing XRT.

**Comment by stsoe** [comment]
@keryell I agree with you and @lforg37 that this PR should not have been merged if it has known bugs / deficiencies / debt.  Your reviews are much appreciated so please keep doing this.   The emulation code is in a very poor state and apparently difficult to fix without a complete rewrite.

**Comment by chvamshi-xilinx** [comment]
Hi @keryell / @lforg37 , I really didn't mean to ignore your suggestions. I have merged it as PR filer created a CR to address your comments and asked us to merge this change.
As Soren mentioned, your review and feedback is very useful . Please keep doing this.
if things are blocked due to any PR, We merge the change and suggests PR filer to create CR to track the comments.

File: src/runtime_src/core/include/xcl_api_macros.h [hunk 5]
```diff
@@ -18,43 +18,54 @@
...
 #include "xcl_macros.h"
 
+#define SCOPE_GUARD_MUTEX() \
+if ( sock->server_started == false ) { std::cerr<<"\n socket communication is not possible now!"; exit(0);  }\
+std::lock_guard<std::mutex> socketlk{mtx}; 
+
 
 #define AQUIRE_MUTEX() \
-mtx.lock(); \
+if ( sock->server_started == false ) { std::cerr<<"\n socket communication is not possible now!"; exit(0);  }\
```
Comment (keryell) [inline src/runtime_src/core/include/xcl_api_macros.h:27]: Why `exit(0)` here ?
Is this to celebrate some happiness with such a success?

**Review by keryell — COMMENTED** [review]
It is unclear this solves anything.
Perhaps the race condition did not append with the new code by chance?

**Pull request merged**